[invoke_subgraph] Support pending unbacked symint

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #149298
* __->__ #149297
* #149296

The "PendingUnbackedSymbolNotFound" error is when an unbacked symbol is created within a piece of code, but this symbol never appears in any of the outputs. I believe the original intention is to help catch incorrectly written meta kernels, where users might've unintentionally created an unbacked symbol but never used it anywhere, but in our case this is intentional. An example is the following test case:

```python
    def test_pending_unbacked(self):
        class M(torch.nn.Module):
            @mark_compile_region
            def gn(self, x):
                u = x[0].item()
                return x * u

            def forward(self, x):
                for _ in range(4):
                    x = self.gn(x)
                return x
        
        torch._dynamo.config.capture_scalar_outputs = True
        torch.compile(M())(torch.randn(8))
```

This fails with the error:
```
torch._dynamo.exc.InternalTorchDynamoError: PendingUnbackedSymbolNotFound: Pending unbacked symbols {zuf1} not in returned outputs (FakeTensor(..., size=(8,)),) .
```

In this case, creating the unbacked symbol is intentional, so we can bypass this using `fake_mode.shape_env.ignore_fresh_unbakced_symbols()`.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

Differential Revision: [D71298926](https://our.internmc.facebook.com/intern/diff/D71298926)

angelayi committed:
[invoke_subgraph] Support pending unbacked symint

[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Support pending unbacked symint"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71298926).

angelayi committed:
Update on "[invoke_subgraph] Support pending unbacked symint"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71298926](https://our.internmc.facebook.com/intern/diff/D71298926)

[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Support pending unbacked symint"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71298926](https://our.internmc.facebook.com/intern/diff/D71298926)

[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Support pending unbacked symint"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71298926](https://our.internmc.facebook.com/intern/diff/D71298926)

[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Support pending unbacked symint"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71298926](https://our.internmc.facebook.com/intern/diff/D71298926)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71298926).

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -3061,6 +3061,7 @@ def call_function(
             for node in body_gmod.graph.nodes
             if node.op == "placeholder"
         ]
+
```
Could we get a summary of what the pending_unbacked problem is?
Also, does the torch.compile version of invoke_subgraph have this issue too?

angelayi replied:
Updated the summary! And yes torch.compile also runs into the same issue if capture_scalar_outputs is set to True.

zou3519 replied:
If you're fixing this for compile too, can you add a compile test too?

angelayi committed:
Update on "[invoke_subgraph] Support pending unbacked symint"


The "PendingUnbackedSymbolNotFound" error is when an unbacked symbol is created within a piece of code, but this symbol never appears in any of the outputs. I believe the original intention is to help catch incorrectly written meta kernels, where users might've unintentionally created an unbacked symbol but never used it anywhere, but in our case this is intentional. An example is the following test case:

```python
    def test_pending_unbacked(self):
        class M(torch.nn.Module):
            mark_compile_region
            def gn(self, x):
                u = x[0].item()
                return x * u

            def forward(self, x):
                for _ in range(4):
                    x = self.gn(x)
                return x
        
        torch._dynamo.config.capture_scalar_outputs = True
        torch.compile(M())(torch.randn(8))
```

This fails with the error:
```
torch._dynamo.exc.InternalTorchDynamoError: PendingUnbackedSymbolNotFound: Pending unbacked symbols {zuf1} not in returned outputs (FakeTensor(..., size=(8,)),) .
```

In this case, creating the unbacked symbol is intentional, so we can bypass this using `fake_mode.shape_env.ignore_fresh_unbakced_symbols()`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71298926](https://our.internmc.facebook.com/intern/diff/D71298926)

[ghstack-poisoned]

angelayi committed:
Update on "[invoke_subgraph] Support pending unbacked symint"


The "PendingUnbackedSymbolNotFound" error is when an unbacked symbol is created within a piece of code, but this symbol never appears in any of the outputs. I believe the original intention is to help catch incorrectly written meta kernels, where users might've unintentionally created an unbacked symbol but never used it anywhere, but in our case this is intentional. An example is the following test case:

```python
    def test_pending_unbacked(self):
        class M(torch.nn.Module):
            mark_compile_region
            def gn(self, x):
                u = x[0].item()
                return x * u

            def forward(self, x):
                for _ in range(4):
                    x = self.gn(x)
                return x
        
        torch._dynamo.config.capture_scalar_outputs = True
        torch.compile(M())(torch.randn(8))
```

This fails with the error:
```
torch._dynamo.exc.InternalTorchDynamoError: PendingUnbackedSymbolNotFound: Pending unbacked symbols {zuf1} not in returned outputs (FakeTensor(..., size=(8,)),) .
```

In this case, creating the unbacked symbol is intentional, so we can bypass this using `fake_mode.shape_env.ignore_fresh_unbakced_symbols()`.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D71298926](https://our.internmc.facebook.com/intern/diff/D71298926)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71298926).

zou3519 commented on torch/_higher_order_ops/invoke_subgraph.py:
```diff
@@ -303,7 +315,7 @@ def _(ctx, subgraph, identifier, operands):
 def _(mode, subgraph, identifier, operands):
     # TODO(anijain2305) - Implement fake tensor caching.
-    with mode:
+    with mode, mode.shape_env.ignore_fresh_unbacked_symbols():
```
I don't know much about pending unbacked, @bobrenjc93 can you confirm that this is just a feature to prevent people from writing incorrect metas?

bobrenjc93 replied:
IIUC this is less about if the symbols occur in the output but rather it's about ensuring that the binding sites are encoded in the fx graph. I'm not too familiar with this mark_compiled_region work. Is this implemented as a HOP? If so this is a common reason to use ignore_fresh_unbacked_symbols. You can see @ydwu4 do something similar for while loops in https://github.com/pytorch/pytorch/pull/142031/files.

zou3519 commented on test/higher_order_ops/test_invoke_subgraph.py:
```diff
@@ -695,6 +695,25 @@ def fn(x):
         self.assertEqual(ref, res)
 
+    def test_pending_unbacked(self):
+        @mark_compile_region
+        def gn(x):
+            u = x[0].item()
+            return x * u
```
I think I am more curious about how you ran into this

angelayi replied:
The deepseek model has a part that creates a mask and then indexes into the mask (creating the unbacked symint) but then the result has a fixed shape that is not dependent on the masked value.

angelayi commented on test/higher_order_ops/test_invoke_subgraph.py:
```diff
@@ -695,6 +695,25 @@ def fn(x):
+        torch._dynamo.mark_dynamic(x, 0)
+        ref = fn(x)
+        torch._dynamo.config.capture_scalar_outputs = True
+        opt_fn = torch.compile(
+            fn, backend="eager", fullgraph=True
+        )  # Inductor fails with cpp compilation error
+        res = opt_fn(x)
```
When compiling with cpu the resulting code is: P1765195315, where the cpp optimized code is missing the value zuf5. This also appears when compiling on gpu, with the logs being P1765200575 and showing the triton kernel is also missing a zuf5
